### DISCUSSION TITLE
An OpenMP issue when hipsycl configured with nvc++ compiler

### DISCUSSION THREAD

**Discussion by ravil-mobile** [AdaptiveCpp/AdaptiveCpp#897] [Q&A]
Hi all, 

I am struggling with the following linking error even though I am compiling code for  `cuda-nvcxx:sm_86` HIPSYCL_TARGETS

```
nvlink error   : Undefined reference to '__kmpc_threadprivate_cached'
```

You can see my `syclcc` configuration below
```
syclcc [hipSYCL compilation driver], Copyright (C) 2018-2022 Aksel Alpay and the hipSYCL project
  hipSYCL version: 0.9.3
  Installation root: /nfs/c3po/home/ga83dit/usr/nvhpc-22-11/hipsycl-0.9.3
  Available runtime backends:
     librt-backend-cuda.so
     librt-backend-omp.so


Full configuration [can be overriden using environment variables or command line arguments]:
    version-major: 0
    version-minor: 9
    version-patch: 3
    plugin-llvm-version-major: 0
    plugin-with-cpu-acceleration: false
    default-clang: (unconfigured)
    default-nvcxx: /nfs/c3po/software/manually-installed/nvhpc_2022_2211/Linux_x86_64/22.11/compilers/bin/nvc++
    default-platform: cuda
    default-cuda-path: /nfs/c3po/software/manually-installed/nvhpc_2022_2211/Linux_x86_64/22.11/cuda
    default-gpu-arch: sm_86
    default-cpu-cxx: /nfs/c3po/software/manually-installed/nvhpc_2022_2211/Linux_x86_64/22.11/compilers/bin/nvc++
    default-rocm-path: /opt/rocm
    default-use-bootstrap-mode: false
    default-is-dryrun: false
    default-use-accelerated-cpu: false
    default-clang-include-path: (unconfigured)
    default-sequential-link-line: -L/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib -lboost_context -lboost_fiber -Wl,-rpath=/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib
    default-sequential-cxx-flags: -I/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-omp-link-line: -L/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib -lboost_context -lboost_fiber -Wl,-rpath=/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib -fopenmp
    default-omp-cxx-flags: -I/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/include -fopenmp -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-rocm-link-line: -Wl,-rpath=$HIPSYCL_ROCM_PATH/lib -Wl,-rpath=$HIPSYCL_ROCM_PATH/hip/lib -L/opt/rocm/lib -L/opt/rocm/hip/lib -lamdhip64
    default-rocm-cxx-flags: -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike -isystem  -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I$HIPSYCL_ROCM_PATH/include -I$HIPSYCL_ROCM_PATH/include --rocm-device-lib-path=$HIPSYCL_ROCM_PATH/amdgcn/bitcode --rocm-path=$HIPSYCL_ROCM_PATH -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__
    default-cuda-link-line: -Wl,-rpath=$HIPSYCL_CUDA_LIB_PATH -L$HIPSYCL_CUDA_LIB_PATH -lcudart
    default-cuda-cxx-flags: -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike
    default-is-explicit-multipass: false
    default-save-temps: false
```

Here is the piece of code which is causing the issue

```c++
  void resampleStateVar(real (*devStateVariableBuffer)[misc::numPaddedPoints]) {
    auto* devStateVariable{this->stateVariable};
    auto* resampleMatrix{this->resampleMatrix};

    constexpr auto dim0 = misc::dimSize<init::resample, 0>();
    constexpr auto dim1 = misc::dimSize<init::resample, 1>();
    static_assert(dim0 == misc::numPaddedPoints);
    static_assert(dim0 >= dim1);

    sycl::nd_range rng{{this->currLayerSize * misc::numPaddedPoints}, {misc::numPaddedPoints}};
    this->queue.submit([&](sycl::handler& cgh) {
      sycl::accessor<real, 1, sycl::access::mode::read_write, sycl::access::target::local>
          deltaStateVar(misc::numPaddedPoints, cgh);

      cgh.parallel_for(rng, [=](sycl::nd_item<1> item) {
        const auto ltsFace = item.get_group().get_group_id(0);
        const auto pointIndex = item.get_local_id(0);

        const auto localStateVariable = devStateVariable[ltsFace][pointIndex];
        deltaStateVar[pointIndex] =
            devStateVariableBuffer[ltsFace][pointIndex] - localStateVariable;
        item.barrier(sycl::access::fence_space::local_space);

        real resampledDeltaStateVar{0.0};
        for (size_t i{0}; i < dim1; ++i) {
          resampledDeltaStateVar += resampleMatrix[pointIndex + i * dim0] * deltaStateVar[i];
        }

        devStateVariable[ltsFace][pointIndex] =
            sycl::max(static_cast<real>(0.0), localStateVariable + resampledDeltaStateVar);
      });
    });
  }
```

The linking problem goes away when I comment the following lines (which uses the local memory)

```c++
/*
        deltaStateVar[pointIndex] =
            devStateVariableBuffer[ltsFace][pointIndex] - localStateVariable;
*/
...
/*
        for (size_t i{0}; i < dim1; ++i) {
          resampledDeltaStateVar += resampleMatrix[pointIndex + i * dim0] * deltaStateVar[i];
        }
*/
```

As far as I understand it doesn't matter for which concrete target I am generating the code. The OpenMP version will be always generated. Meanwhile I am working on creating a small reproducible example. 

It would be great if somebody can suggest something or share some ideas how to circumvent the issue.

Thanks in advance!

**Comment by illuhad** [comment]
Hi,

It always generates a code path to fall back on the CPU in case no GPU is available, but it should only actually use OpenMP pragmas when compiling for the `omp` target.
Can you confirm if you are on 0.9.3 release, or latest develop branch? I think there have been some fixes regarding the OpenMP dependency relatively recently.
On the current `develop` branch, the OpenMP `threadprivate` feature (which is likely causing the issue) is only used when `_OPENMP` is defined, i.e. when actually targeting the `omp` backend:
https://github.com/illuhad/hipSYCL/blob/58675bda6143d2881f6cb9def4cd12932cae07c5/include/hipSYCL/sycl/libkernel/detail/local_memory_allocator.hpp#L167

I would recommend upgrading to the latest `develop` branch if you are using an older version.

Of course, alternatively you could also try to make OpenMP work with nvc++. Probably `nvc++ -fopenmp` does not correctly link the OpenMP runtime library. You could try to manually link with `libomp` or `libgomp` to resolve the issue.

**Reply by ravil-mobile** [reply]
Hi @illuhad, thanks for your quick reply.

> Can you confirm if you are on 0.9.3 release, or latest develop branch? I think there have been some fixes regarding the OpenMP dependency relatively recently.

Yes, sure.

```bash
$ git branch -v
* develop 58675bda Fix LLVM 16 compat. (#893)
```

> Probably nvc++ -fopenmp does not correctly link the OpenMP runtime library.

You're right. `nvc++` uses `-mp` flag for OpenMP. I slightly changed cmake as follows:
  
```diff
git diff CMakeLists.txt
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6328fb1d..8b26a7da 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -234,7 +234,11 @@ if(APPLE)
     set(DEFAULT_BOOST_LIBRARIES "${Boost_CONTEXT_LIBRARY_RELEASE} ${Boost_FIBER_LIBRARY_RELEASE} -Wl,-rpath ${Boost_LIBRARY_DIR}")
   endif()
 else()
-  set(DEFAULT_OMP_FLAG "-fopenmp")
+  if (CMAKE_CXX_COMPILER_ID MATCHES "NVHPC|PGI")
+    set(DEFAULT_OMP_FLAG "-mp")
+  else()
+    set(DEFAULT_OMP_FLAG "-fopenmp")
+  endif()
 endif()
```

<details>
<summary>my current hipSYCL configuration</summary>

```bash
syclcc [hipSYCL compilation driver], Copyright (C) 2018-2022 Aksel Alpay and the hipSYCL project
  hipSYCL version: 0.9.3
  Installation root: /nfs/c3po/home/ga83dit/usr/nvhpc-22-11/hipsycl-0.9.3
  Available runtime backends:
     librt-backend-cuda.so
     librt-backend-omp.so


Full configuration [can be overriden using environment variables or command line arguments]:
    version-major: 0
    version-minor: 9
    version-patch: 3
    plugin-llvm-version-major: 0
    plugin-with-cpu-acceleration: false
    default-clang: (unconfigured)
    default-nvcxx: /nfs/c3po/software/manually-installed/nvhpc_2022_2211/Linux_x86_64/22.11/compilers/bin/nvc++
    default-platform: cuda
    default-cuda-path: /nfs/c3po/software/manually-installed/nvhpc_2022_2211/Linux_x86_64/22.11/cuda
    default-gpu-arch: sm_86
    default-cpu-cxx: /nfs/c3po/software/manually-installed/nvhpc_2022_2211/Linux_x86_64/22.11/compilers/bin/nvc++
    default-rocm-path: /opt/rocm
    default-use-bootstrap-mode: false
    default-is-dryrun: false
    default-use-accelerated-cpu: false
    default-clang-include-path: (unconfigured)
    default-sequential-link-line: -L/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib -lboost_context -lboost_fiber -Wl,-rpath=/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib
    default-sequential-cxx-flags: -I/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/include -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-omp-link-line: -L/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib -lboost_context -lboost_fiber -Wl,-rpath=/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/lib -mp
    default-omp-cxx-flags: -I/u/home/ga83dit/usr/nvhpc-22-11/boost-1.69.0/include -mp -D_ENABLE_EXTENDED_ALIGNED_STORAGE
    default-rocm-link-line: -Wl,-rpath=$HIPSYCL_ROCM_PATH/lib -Wl,-rpath=$HIPSYCL_ROCM_PATH/hip/lib -L/opt/rocm/lib -L/opt/rocm/hip/lib -lamdhip64
    default-rocm-cxx-flags: -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike -isystem  -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -I$HIPSYCL_ROCM_PATH/include -I$HIPSYCL_ROCM_PATH/include --rocm-device-lib-path=$HIPSYCL_ROCM_PATH/amdgcn/bitcode --rocm-path=$HIPSYCL_ROCM_PATH -fhip-new-launch-api -mllvm -amdgpu-early-inline-all=true -mllvm -amdgpu-function-calls=false -D__HIP_ROCclr__
    default-cuda-link-line: -Wl,-rpath=$HIPSYCL_CUDA_LIB_PATH -L$HIPSYCL_CUDA_LIB_PATH -lcudart
    default-cuda-cxx-flags: -U__FLOAT128__ -U__SIZEOF_FLOAT128__ -isystem $HIPSYCL_PATH/include/hipSYCL/std/hiplike
    default-is-explicit-multipass: false
    default-save-temps: false

asd
```

</details>

But it didn't help to resolve the linking problem.

> the OpenMP threadprivate feature (which is likely causing the issue) is only used when _OPENMP is defined, i.e. when actually targeting the omp backend

You're correct. The `threadprivate` clause is causing the issue. I did the following to circumvent the linking issue

```bash
$ sycl_path=$(dirname $(dirname $(which syclcc)))
$ sycl_allocator_file="$sycl_path/include/hipSYCL/sycl/libkernel/detail/local_memory_allocator.hpp"
$ sed -i 's/\(#pragma omp threadprivate\)/\/\/\1/g' ${sycl_allocator_file}
```
After that I managed to compile my code.

`_OPENMP` is defined by the compiler when the corresponding openmp flag is set (e.g., `-fopenmp`, `-mp`). The `threadprivate` clause is defined in a header file which is included into the installation folder. Therefore, I guess the clause becomes a part of my code caused by a chain of includes starting from `#include <CL/sycl.hpp>`. 

My code uses both OpenMP and SYCL and thus I am compiling the code with the openmp flag which forces the compiler to include `_OPENMP` definition.

My point is that maybe `hipSYCL` needs to rely on some other custom C-macro rather than on `_OPENMP`. What do you think? However, it does not explain my linking issue. The problem is strange because everything is linking properly when I am compiling my a small reproducible example.

**Reply by fodinabor** [reply]
Hmm.. according to new OpenMP standards (5.1 & 5.2) , C++11's `thread_local` should have the same semantics as `#pragma omp threadprivate`. Not sure if that is already the case with all compilers (including nvc++), but might be worth testing if marking the variables in question `thread_local` and removing the OpenMP directive 1. solves the linking issue and 2. does not alter the semantics.

A simple way to test whether the semantics are preserved should be to enqueue a kernel using a small local memory accessor (so that it fits `_static_local_mem`, just a single element?) with at least 2 work-groups on CPU and print the local memory ptrs for the first work-item. They should be equivalent if no `thread_local` modifier or `threadprivate`  directive is present, but be equal if either is used.

**Reply by illuhad** [reply]
>  You're right. nvc++ uses -mp flag for OpenMP. I slightly changed cmake as follows:

A basic OpenMP program also seems to work correctly with `-fopenmp`, so nvc++ seems to support `fopenmp` for compatibility: https://godbolt.org/z/j6KbePWsY
So, this also does not explain the issue.

> My code uses both OpenMP and SYCL and thus I am compiling the code with the openmp flag which forces the compiler to include _OPENMP definition.

Ok, but if something else in your code enables OpenMP, the compiler should still be handling OpenMP pragmas correctly, and link as required for OpenMP. So this still smells like some nvc++ OpenMP bug to me. Does nvc++ not support threadprivate?
For hipSYCL, I don't think it should matter whether hipSYCL enables OpenMP or some external part of the software infrastructure does.

> The problem is strange because everything is linking properly when I am compiling my a small reproducible example.

Also when using the threadprivate pragma?

**Reply by ravil-mobile** [reply]
Hi @illuhad and @fodinabor,

Thanks for your replies and for sharing your ideas.

> Also when using the threadprivate pragma?

I have two dedicated examples i.e., 1) using a piece of code that I shared with you in my first message (with mocked device memory) and 2) nvc++ with `omp threadprivate`. I can compile both examples without any issue.

> So this still smells like some nvc++ OpenMP bug to me

Yes, it smells like that. 

I will keep digging. As soon as I find how to reproduce the problem with a small example I will share the code with you (of course, if you're interested in).

**Reply by illuhad** [reply]
>  and 2) nvc++ with omp threadprivate. I can compile both examples without any issue.

It might only occur perhaps when mixing CUDA and OpenMP in the same source file which happens in hipSYCL, because nvc++ needs to compile kernels both for host and device.
So maybe also put both a CUDA kernel in there, and in a different part of the code an `omp threadprivate`, and compile with `nvc++ -cuda -fopenmp`.

> I will keep digging. As soon as I find how to reproduce the problem with a small example I will share the code with you (of course, if you're interested in).

Of course I am interested in that :)

**Comment by patrickatkinson** [comment]
Hi @illuhad,

Ravil pointed me towards this issue, and I’ve had a chance to investigate. This seems like a problem related to combining the `if target` construct with `omp threadprivate`, and it only seems to happen with `-O2` or `-O3`, but not with `-O1`. For me this problem seemed to affect any usage of local memory in OpenSYCL when compiling with OpenMP and O2 or O3.  

I have a MVE that seems to reproduce the linking error originating from the `local_memory` class that can be compiled without OpenSYCL:

```
#include <stdio.h>

#include <nv/target>

class host_local_memory_test {
public:
  static char* get_ptr() { return static_local_mem; }

private:
  static char static_local_mem[1024];
#pragma omp threadprivate(static_local_mem)
};
char host_local_memory_test::static_local_mem[1024] = {0};

class local_memory_test {
public:
  template <class T>
  __host__ __device__ static T* get_ptr(const size_t addr)
  {
    if target (nv::target::is_host) {
      return reinterpret_cast<T*>(host_local_memory_test::get_ptr() + addr);
    }
    else if target (nv::target::is_device) {
      return nullptr;
    }
  }
};

__global__ void kernel()
{
  char* ptr = local_memory_test::get_ptr<char>((size_t)0);
  printf("ptr = %p\n", ptr);
}

int main()
{
  kernel<<<1, 1>>>();
  cudaDeviceSynchronize();
  return 0;
}
```

I have managed to reproduce with NVHPC 22.11 and 23.1; however, the issue is now fixed in NVHPC 23.3, which was released last week. 

Hopefully using NVHPC 23.3 fixes the issue for both of you. If not, let me know and I can take a further look.

**Reply by illuhad** [reply]
That's great news, thanks for the feedback!
